Add back frame visualization in Meshcat

hjsuh94 opened the pull request:
The `frames_to_draw` argument in `ConnectMeshcatVisualizer` got accidentally deprecated in #14292 (this seems to be unintentional.) 

This is a tiny PR to add back the feature with the model inspector syntax.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/14979)
<!-- Reviewable:end -->

RussTedrake commented:
i've not reviewed this (yet), but FTR the original change to queryobject was actually done, in part, to update the frames_to_draw api: https://github.com/RobotLocomotion/drake/issues/13934 .

hjsuh94 commented:
I've updated the api to accept a list of `pydrake.geometry.FrameId`s, as opposed to the ugly(?) dictionary we had before.

This requires the user to query the id from their applications, but this will handle the concerns in #13934. I've updated the test files as well to match the new api. 

+@RussTedrake for review please! Maybe @pangtao22 could give a pass as well?

RussTedrake reviewed (commented):
Thanks @hjsuh94 !

> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14979#-:-MZhCGhn2hBBLW3-2-95:b-om3x66)* status: 2 unresolved discussions, LGTM missing from assignee RussTedrake(platform), needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @hjsuh94)

___
*[bindings/pydrake/systems/meshcat_visualizer.py, line 168 at r2](https://reviewable.io/reviews/robotlocomotion/drake/14979#-MZh8_ev4sbRMHOvlfnP:-MZh8_ev4sbRMHOvlfnQ:b2vhodu) ([raw file](https://github.com/robotlocomotion/drake/blob/2b5cadfecac56d8ee143a517226117218778d80a/bindings/pydrake/systems/meshcat_visualizer.py#L168)):*
> ```Python
>             frames_to_draw: a list or array containing pydrake.geometry.FrameId
>                 for which body frames to draw. Users may obtain the ids from
>                 plant using GetBodyFrameIdIfExists.
> ```

Nice.  "For multibody frames, users may obtain.."   (the point is frames in scenegraph need not come from multibody, which is why this generalization was important).  And we should make sure to say the FrameID must be declared with the SceneGraph.  see note below.
___
*[bindings/pydrake/systems/meshcat_visualizer.py, line 418 at r2](https://reviewable.io/reviews/robotlocomotion/drake/14979#-MZh8ukN9QbT-eUhVJV3:-MZh8ukN9QbT-eUhVJV4:b-vgsi1r) ([raw file](https://github.com/robotlocomotion/drake/blob/2b5cadfecac56d8ee143a517226117218778d80a/bindings/pydrake/systems/meshcat_visualizer.py#L418)):*
> ```Python
>                     AddTriad(
>                         self.vis,
>                         name="frame",
> ```

should use the frame name here (obtained from scenegraph, as @seancurtis-tri mentioned in that issue)?
___
*[bindings/pydrake/systems/meshcat_visualizer.py, line 435 at r2](https://reviewable.io/reviews/robotlocomotion/drake/14979#-MZh93WE6tHivH5Ny6QL:-MZh93WE6tHivH5Ny6QM:b-kzwapb) ([raw file](https://github.com/robotlocomotion/drake/blob/2b5cadfecac56d8ee143a517226117218778d80a/bindings/pydrake/systems/meshcat_visualizer.py#L435)):*
> ```Python
>         for frame_id in self.frames_to_draw:
>             if (frame_id not in inspector.all_frame_ids()):
>                 warnings.warn(f"Non-existent frame {frame_id} ignored")
> ```

btw -- nice.
___
*[bindings/pydrake/systems/test/meshcat_visualizer_test.py, line 127 at r2](https://reviewable.io/reviews/robotlocomotion/drake/14979#-MZh9JXE--q4aAIAxB7u:-MZh9JXE--q4aAIAxB7v:bkaksa3) ([raw file](https://github.com/robotlocomotion/drake/blob/2b5cadfecac56d8ee143a517226117218778d80a/bindings/pydrake/systems/test/meshcat_visualizer_test.py#L127)):*
> ```Python
>                 # Make sure the frames to visualize exists.
>                 kuka.GetBodyFrameIdOrThrow(
>                     kuka.GetFrameByName(frame_name).body().index()
> ```

I think this is correct, but is admittedly not pretty!  I think this hits at the heart of a deeper problem with the pipeline:  not all [Frames](https://drake.mit.edu/doxygen_cxx/classdrake_1_1multibody_1_1_frame.html)  from a `MultibodyPlant` are registered with the `SceneGraph` (only the body frames are).  This sure seems like a defect on the MBP <-> SG relationship to me.  I looked for a discussion about it, but perhaps @seancurtis-tri can help me understand?   Should I open an issue on it?

To be clear, it is not a defect with this PR and I've set the comment status to "informing".  The only impact on this PR is that we should make the distinction (SceneGraph frames, not MBP frames) as clear as possible to the user.


<!-- Sent from Reviewable.io -->

hjsuh94 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14979)* status: 2 unresolved discussions, LGTM missing from assignee RussTedrake(platform), needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @RussTedrake and @SeanCurtis-TRI)

___
*[bindings/pydrake/systems/meshcat_visualizer.py, line 168 at r2](https://reviewable.io/reviews/robotlocomotion/drake/14979#-MZh8_ev4sbRMHOvlfnP:-MZhIOqKDijPbETgTHgy:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/2b5cadfecac56d8ee143a517226117218778d80a/bindings/pydrake/systems/meshcat_visualizer.py#L168)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

Nice.  "For multibody frames, users may obtain.."   (the point is frames in scenegraph need not come from multibody, which is why this generalization was important).  And we should make sure to say the FrameID must be declared with the SceneGraph.  see note below.
</blockquote></details>

Done.
___
*[bindings/pydrake/systems/meshcat_visualizer.py, line 418 at r2](https://reviewable.io/reviews/robotlocomotion/drake/14979#-MZh8ukN9QbT-eUhVJV3:-MZhISr1Dw6D8RprxrZc:b-s5tguo) ([raw file](https://github.com/robotlocomotion/drake/blob/2b5cadfecac56d8ee143a517226117218778d80a/bindings/pydrake/systems/meshcat_visualizer.py#L418)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

should use the frame name here (obtained from scenegraph, as @seancurtis-tri mentioned in that issue)?
</blockquote></details>

Done. Good catch! 
___
*[bindings/pydrake/systems/test/meshcat_visualizer_test.py, line 127 at r2](https://reviewable.io/reviews/robotlocomotion/drake/14979#-MZh9JXE--q4aAIAxB7u:-MZhIZkIEDePXkrsL7As:b-ekadtw) ([raw file](https://github.com/robotlocomotion/drake/blob/2b5cadfecac56d8ee143a517226117218778d80a/bindings/pydrake/systems/test/meshcat_visualizer_test.py#L127)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

I think this is correct, but is admittedly not pretty!  I think this hits at the heart of a deeper problem with the pipeline:  not all [Frames](https://drake.mit.edu/doxygen_cxx/classdrake_1_1multibody_1_1_frame.html)  from a `MultibodyPlant` are registered with the `SceneGraph` (only the body frames are).  This sure seems like a defect on the MBP <-> SG relationship to me.  I looked for a discussion about it, but perhaps @seancurtis-tri can help me understand?   Should I open an issue on it?

To be clear, it is not a defect with this PR and I've set the comment status to "informing".  The only impact on this PR is that we should make the distinction (SceneGraph frames, not MBP frames) as clear as possible to the user.
</blockquote></details>

I agree with this comment. Two issues here:
1. nit: as a user, I did briefly struggle to find the right syntax for this (which is what motivated me to include `GetBodyFrameIdorThrow` in the documentation for `frames_to_draw`) - it is certainly a mouthful!
2. To illustrate your second point, I couldn't add a `Frame`tag in my sdf file and visualize it using `frames_to_draw` because of the exact reasons you mentioned. (The workaround was to create a dummy link just for visualization, which works but is not ideal).


<!-- Sent from Reviewable.io -->

RussTedrake reviewed (commented):
> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14979)* status: LGTM missing from assignee RussTedrake(platform), needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @hjsuh94)

___
*[bindings/pydrake/systems/test/meshcat_visualizer_test.py, line 127 at r2](https://reviewable.io/reviews/robotlocomotion/drake/14979#-MZh9JXE--q4aAIAxB7u:-MZhYBy7Bie66E9qquzM:b-wdxw1f) ([raw file](https://github.com/robotlocomotion/drake/blob/2b5cadfecac56d8ee143a517226117218778d80a/bindings/pydrake/systems/test/meshcat_visualizer_test.py#L127)):*
<details><summary><i>Previously, hjsuh94 (Hyung Ju Suh (Terry)) wrote…</i></summary><blockquote>

I agree with this comment. Two issues here:
1. nit: as a user, I did briefly struggle to find the right syntax for this (which is what motivated me to include `GetBodyFrameIdorThrow` in the documentation for `frames_to_draw`) - it is certainly a mouthful!
2. To illustrate your second point, I couldn't add a `Frame`tag in my sdf file and visualize it using `frames_to_draw` because of the exact reasons you mentioned. (The workaround was to create a dummy link just for visualization, which works but is not ideal).
</blockquote></details>

Alternatively, you could actually support the multibody::Frame explicitly here, by writing in this method the logic that goes back up to the Body, connects to the BodyFrame in SceneGraph, and adds the triad at the corresponding relative transform from that body frame here.  That would make the user experience better.  wdyt?


<!-- Sent from Reviewable.io -->

RussTedrake reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>
(we can add the support for multibody:Frame here if you're willing, or in a future PR)
+@seancurtis-tri for platform review, please.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14979#-:-MZhoJV97LX7QqBYAIHQ:bfahsos)* status: LGTM missing from assignee SeanCurtis-TRI(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @SeanCurtis-TRI)



<!-- Sent from Reviewable.io -->

hjsuh94 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14979)* status: LGTM missing from assignee SeanCurtis-TRI(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @SeanCurtis-TRI)

___
*[bindings/pydrake/systems/test/meshcat_visualizer_test.py, line 127 at r2](https://reviewable.io/reviews/robotlocomotion/drake/14979#-MZh9JXE--q4aAIAxB7u:-MZhnw658af7V0SlSZBd:bijez3b) ([raw file](https://github.com/robotlocomotion/drake/blob/2b5cadfecac56d8ee143a517226117218778d80a/bindings/pydrake/systems/test/meshcat_visualizer_test.py#L127)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

Alternatively, you could actually support the multibody::Frame explicitly here, by writing in this method the logic that goes back up to the Body, connects to the BodyFrame in SceneGraph, and adds the triad at the corresponding relative transform from that body frame here.  That would make the user experience better.  wdyt?
</blockquote></details>

Certainly not difficult to do, but just backing up to think about the high level interface....what do you think is the right thing to do?

1. Accept only a list of geometry::frame_id s
2. Accept only a list of multibody::Frame s 
3. Accept both and detect which one it is to handle them separately. 

First one is how the method is intended to be used but has bad user interface, second one also feels unsatisfying because some users might actually want to pass scene graph ids, and the third one feels ad-hoc.

If I were to choose I’d go with the third one though :) (just saw your lgtm as I was typing this up so we can worry about this later too) 


<!-- Sent from Reviewable.io -->

hjsuh94 commented:
I tried giving this a pass tonight but turns out to be an uglier problem than I thought it would be. I think we can try to merge this one as it is.

For record to myself, here are the places where it gets complicated:
1. Knowing which `FrameId` the `Frames` belong to is required to get the body frame names from `inspector` (to pass onto Meshcat), but this requires knowledge of which `plant` the `Frames` belong to. This will need to be an additional argument to `ConnectMeshcatVisualizer`, but is unsatisfying since this is the only place `plant` will be used, and the visualizer should ideally be only require the `scene_graph`. 
2. Even if we manage to bypass adding a dummy link for `scene_graph`, we would need to declare one for Meshcat. I don't think this is a roadblock however, if the first problem is solved.

pangtao22 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14979)* status: LGTM missing from assignee SeanCurtis-TRI(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @SeanCurtis-TRI)

___
*[bindings/pydrake/systems/test/meshcat_visualizer_test.py, line 127 at r2](https://reviewable.io/reviews/robotlocomotion/drake/14979#-MZh9JXE--q4aAIAxB7u:-MZjxu6d0XR4xl4MS27-:bbxq31) ([raw file](https://github.com/robotlocomotion/drake/blob/2b5cadfecac56d8ee143a517226117218778d80a/bindings/pydrake/systems/test/meshcat_visualizer_test.py#L127)):*
<details><summary><i>Previously, hjsuh94 (Hyung Ju Suh (Terry)) wrote…</i></summary><blockquote>

Certainly not difficult to do, but just backing up to think about the high level interface....what do you think is the right thing to do?

1. Accept only a list of geometry::frame_id s
2. Accept only a list of multibody::Frame s 
3. Accept both and detect which one it is to handle them separately. 

First one is how the method is intended to be used but has bad user interface, second one also feels unsatisfying because some users might actually want to pass scene graph ids, and the third one feels ad-hoc.

If I were to choose I’d go with the third one though :) (just saw your lgtm as I was typing this up so we can worry about this later too) 
</blockquote></details>

Since we're limited to drawing body frames here, we could also use the following to make the incantation a bit shorter. 
 ```                 
kuka.GetBodyFrameIdOrThrow(kuka.GetBodyByName(frame_name).index())
 ```


<!-- Sent from Reviewable.io -->

pangtao22 reviewed (commented):
Another way to draw a frame that is not a multibody::Frame in meshcat is to call `AddTriad` directly, which doesn't need to go through SceneGraph or MultibodyPlant. We can also set relative transform between the added frame and its base frame using meshcat's API. 

This is clearly not as ideal as having MeshcatVisualier handle it, but it seems like the discrepancy between multibody::Frame and geometry::Frame needs to be sorted out first...

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14979#-:-MZjzBbQ4km_YfiSLfer:b7oswht)* status: LGTM missing from assignee SeanCurtis-TRI(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @SeanCurtis-TRI)



<!-- Sent from Reviewable.io -->

hjsuh94 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14979)* status: LGTM missing from assignee SeanCurtis-TRI(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @RussTedrake and @SeanCurtis-TRI)

___
*[bindings/pydrake/systems/test/meshcat_visualizer_test.py, line 127 at r2](https://reviewable.io/reviews/robotlocomotion/drake/14979#-MZh9JXE--q4aAIAxB7u:-MZk21Rj9ZXPruf0Pwhf:bj4bm4n) ([raw file](https://github.com/robotlocomotion/drake/blob/2b5cadfecac56d8ee143a517226117218778d80a/bindings/pydrake/systems/test/meshcat_visualizer_test.py#L127)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary><blockquote>

Since we're limited to drawing body frames here, we could also use the following to make the incantation a bit shorter. 
 ```                 
kuka.GetBodyFrameIdOrThrow(kuka.GetBodyByName(frame_name).index())
 ```

</blockquote></details>

Done. Thanks Pang! 


<!-- Sent from Reviewable.io -->

RussTedrake reviewed (commented):
> Reviewed 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14979)* status: LGTM missing from assignee SeanCurtis-TRI(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @SeanCurtis-TRI)

___
*[bindings/pydrake/systems/test/meshcat_visualizer_test.py, line 127 at r2](https://reviewable.io/reviews/robotlocomotion/drake/14979#-MZh9JXE--q4aAIAxB7u:-MZlc_Hy7FgshTh8pTIW:b-zhsqkd) ([raw file](https://github.com/robotlocomotion/drake/blob/2b5cadfecac56d8ee143a517226117218778d80a/bindings/pydrake/systems/test/meshcat_visualizer_test.py#L127)):*
<details><summary><i>Previously, hjsuh94 (Hyung Ju Suh (Terry)) wrote…</i></summary><blockquote>

Done. Thanks Pang! 
</blockquote></details>

can you update this in the method documentation, too?


<!-- Sent from Reviewable.io -->

SeanCurtis-TRI reviewed (commented):
<img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> With one non-binding suggestion. If you're interested, I'll let you do it. Otherwise, feel free to merge.

> Reviewed 1 of 1 files at r3, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14979#-:-MZmtdC76hPCU_HNhF8r:btkmgxk)* status: commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @hjsuh94)

___
*[bindings/pydrake/systems/meshcat_visualizer.py, line 435 at r2](https://reviewable.io/reviews/robotlocomotion/drake/14979#-MZh93WE6tHivH5Ny6QL:-MZmuSBW52o7Tb-_yXhT:boljekf) ([raw file](https://github.com/robotlocomotion/drake/blob/2b5cadfecac56d8ee143a517226117218778d80a/bindings/pydrake/systems/meshcat_visualizer.py#L435)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

btw -- nice.
</blockquote></details>

BTW As an optimization:

You might consider converting the array/list of frame ids into a set when you start and then pop a frame id every time you add a visualization. This will:
  - Speed up the "is this frame in my frames to draw" query
  - Reduce this query to just listing all the remaining frames.
___
*[bindings/pydrake/systems/test/meshcat_visualizer_test.py, line 127 at r2](https://reviewable.io/reviews/robotlocomotion/drake/14979#-MZh9JXE--q4aAIAxB7u:-MZmv2Z190LwvfRxVZk_:bwgxuhi) ([raw file](https://github.com/robotlocomotion/drake/blob/2b5cadfecac56d8ee143a517226117218778d80a/bindings/pydrake/systems/test/meshcat_visualizer_test.py#L127)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

can you update this in the method documentation, too?
</blockquote></details>

There is no reason that MBP couldn't register a `geometry::FrameId` with `SceneGraph` for every non-body frame. It just currently doesn't do it. 

The historical reason is pretty straightforward. Frames are there to move geometries. If we don't have geometries, what's the point of having a frame? This particularly application is the answer to that. I'd actually like to visualize frames!

An issue would certainly be appropriate.


<!-- Sent from Reviewable.io -->

hjsuh94 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14979)* status: commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @RussTedrake and @SeanCurtis-TRI)

___
*[bindings/pydrake/systems/meshcat_visualizer.py, line 435 at r2](https://reviewable.io/reviews/robotlocomotion/drake/14979#-MZh93WE6tHivH5Ny6QL:-MZnAchz41vx-WSktBNI:bhgw1q3) ([raw file](https://github.com/robotlocomotion/drake/blob/2b5cadfecac56d8ee143a517226117218778d80a/bindings/pydrake/systems/meshcat_visualizer.py#L435)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW As an optimization:

You might consider converting the array/list of frame ids into a set when you start and then pop a frame id every time you add a visualization. This will:
  - Speed up the "is this frame in my frames to draw" query
  - Reduce this query to just listing all the remaining frames.

</blockquote></details>

Done. It's a great suggestion!
___
*[bindings/pydrake/systems/test/meshcat_visualizer_test.py, line 127 at r2](https://reviewable.io/reviews/robotlocomotion/drake/14979#-MZh9JXE--q4aAIAxB7u:-MZnAhrg8Ao0D9zyxOrL:b-tmbv88) ([raw file](https://github.com/robotlocomotion/drake/blob/2b5cadfecac56d8ee143a517226117218778d80a/bindings/pydrake/systems/test/meshcat_visualizer_test.py#L127)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

There is no reason that MBP couldn't register a `geometry::FrameId` with `SceneGraph` for every non-body frame. It just currently doesn't do it. 

The historical reason is pretty straightforward. Frames are there to move geometries. If we don't have geometries, what's the point of having a frame? This particularly application is the answer to that. I'd actually like to visualize frames!

An issue would certainly be appropriate.
</blockquote></details>

Updated the method document to say that only body frames are currently supported. 


<!-- Sent from Reviewable.io -->

jamiesnape commented:
@drake-jenkins-bot linux-bionic-gcc-bazel-experimental-debug please

SeanCurtis-TRI reviewed (commented):
> Reviewed 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14979)* status: :shipit: complete! all discussions resolved, LGTM from assignees SeanCurtis-TRI(platform),RussTedrake(platform) (waiting on @hjsuh94)



<!-- Sent from Reviewable.io -->

hjsuh94 commented:
Ready to merge!
